unliftio/test/UnliftIO/ExceptionSpec.hs

```diff
@@ -1,10 +1,77 @@
 module UnliftIO.ExceptionSpec (spec) where
 
+import Control.Monad (void)
+import Data.Bifunctor (first)
 import Test.Hspec
 import UnliftIO
+import UnliftIO.Concurrent (threadDelay)
 
 spec :: Spec
 spec = do
+  let -- The callback will run in a thread that gets cancelled immediately,
+      -- then get Exception2 thrown synchronously after 1 second.
+      withAsyncExceptionThrown :: (IO a -> IO b) -> IO b
+      withAsyncExceptionThrown f = do
+        var <- newEmptyMVar
+        a <- async $ f $ do
+          putMVar var ()
```

Comment:
Thank you for switching to a baton approach for this!

Reply:
I always had it like this ;) I just tried to simplify it in the convo so that it was clearer what the concern was (even if there was a bit of a race condition)